src/extension.ts

```diff
@@ -194,20 +206,95 @@ function config(section: string, scope?: vscode.ConfigurationScope) {
 	return vscode.workspace.getConfiguration(section, scope);
 }
 
-function workspaceFolder(folder: string) {
-	return vscode.workspace.getWorkspaceFolder(vscode.Uri.parse(folder));
+function getWorkspaceFolder(uri: string): vscode.WorkspaceFolder {
+	return vscode.workspace.getWorkspaceFolder(vscode.Uri.parse(uri));
 }
 
-function folderNames(folders: readonly vscode.WorkspaceFolder[] = vscode.workspace.workspaceFolders): string[] {
-	if (!folders) {
-		return [];
+function getDocumentClient(document: vscode.Uri) {
+	return clients.get(clientName(document.toString()));
+}
+
+function getFolderName(folder: vscode.WorkspaceFolder): string {
+	let result = folder.uri.toString();
+	if (result.charAt(result.length - 1) !== '/') {
+		result = result + '/';
 	}
-	return folders.map(folder => {
-		let result = folder.uri.toString();
-		if (result.charAt(result.length - 1) !== '/') {
-			result = result + '/';
+	return result;
+}
+
+function sortedWorkspaceFolders() {
+	const workspaceFolders = vscode.workspace.workspaceFolders;
+	if (workspaceFolders) {
+		return vscode.workspace.workspaceFolders.map(f => getFolderName(f)).sort(
+			(a, b) => {
+				return a.length - b.length;
+			});
+	}
+	return [];
+}
+
+function clientName(folderName: string, workspaceFolders: readonly string[] = sortedWorkspaceFolders()) {
+	const outerFolder = workspaceFolders.find(element => folderName.startsWith(element));
+	// If this folder isn't nested, the found item will be itself
+	if (outerFolder && (outerFolder !== folderName)) {
+		folderName = getFolderName(getWorkspaceFolder(outerFolder));
+	}
+	return folderName;
+}
+
+function folderNames(folders: readonly vscode.WorkspaceFolder[] = vscode.workspace.workspaceFolders): string[] {
+	const result = [];
+	// Sort workspace folders so that outer folders (shorter path) go before inner ones
+	const workspaceFolders = sortedWorkspaceFolders();
+	if (folders && workspaceFolders) {
+		const folderNames = folders.map(f => getFolderName(f));
+		for (let name of folderNames) {
+			name = clientName(name, workspaceFolders);
+			result.push(name);
 		}
-		return result;
+	}
+
+	return result;
+}
+
+function execWorkspaceCommand(client: LanguageClient, params: ExecuteCommandParams): Promise<any> {
+	return client.sendRequest(ExecuteCommandRequest.type, params);
+}
+
+interface rootModule {
+	uri: string
+}
+
+interface rootModuleResponse  {
+	rootModules: rootModule[],
+	needsInit: boolean
+}
+
+async function rootModulesCommand(client: LanguageClient, documentUri: string): Promise<any> {
+	const requestParams: ExecuteCommandParams = { command: "rootmodules", arguments: [`uri=${documentUri}`] };
+	return execWorkspaceCommand(client, requestParams);
+}
+
+async function rootModules(client: LanguageClient, documentUri: string) {
+	return new Promise<rootModuleResponse>((resolve, reject) => {
```

Comment:
I would like to avoid bringing in `async.js`. Perhaps rewrite function to include retry loop, async/await should be it relatively simple

```javascript
async function rootModules(client: LanguageClient, documentUri: string) {
    let done = false;
    for (let i = 0; i < 2 && !done; i++) {
        const resp = await rootModulesCommand(client, documentUri);
        await sleep(100); // move that sleep helper from the tests to somewhere more common and import it
        done = resp.doneLoading;
    }
    if (!done) {
        throw new Error(`Unable to load root modules for ${documentUri}`);
    }
    return { rootModules: resp.rootModules, needsInit: resp.rootModules.length === 0 };
}
```

No need to `try/catch` even, any error in `rootModulesCommand` will automatically bubble up (one of the best parts of `async/await`)

Reply:
That was what I started with, but it felt clunky (not necessarily a problem!). Is there a specific reason you want to avoid adding the dependency, or just to keep it simple?